### PR TITLE
koord-manager: revise node resource framework

### LINKED ISSUE #1060: [proposal] noderesource controller supports resource calculator extenders
**What is your proposal**:

The node-resource controller of the koord-manager should allow extensions for resource calculation.

**Why is this needed**:

The resource calculation algorithm in the node-resource controller should be extensible. For instance, a non-overcommit extender can be used for some nodes' batch resource calculation, while the other nodes use the default.

**Is there a suggested solution, if so, please add it**:

Add an extender framework in the node-resource controller.

### PR DESCRIPTION
**Pull request by saintube**
### Ⅰ. Describe what this PR does

Revise node resource framework:

1. Add a sequence of the plugins' execution by the registered order.
2. Make the degraded resource calculation private in each ResourceCalculatePlugin.
3. Use structural logs in the node-resource controller.

### Ⅱ. Does this pull request fix one issue?

fixes #1060

### V. Checklist

- [X] I have written necessary docs and comments
- [X] I have added necessary unit tests and integration tests
- [X] All checks passed in `make test`

### CODE REVIEW

File: pkg/slo-controller/noderesource/noderesource_controller.go [hunk 1]
```diff
@@ -86,21 +85,13 @@ func (r *NodeResourceReconciler) Reconcile(ctx context.Context, req ctrl.Request
...
-
-	if r.isDegradeNeeded(nodeMetric, node) {
-		klog.Warningf("node %v need degradation, reset BE resource", req.Name)
-		if err := r.resetNodeResource(node, "degrade node resource because of abnormal nodeMetric, reason: "+degradeByKoordController); err != nil {
+		if !errors.IsNotFound(err) {
+			klog.ErrorS(err, "failed to get nodeMetric", "node", req.Name)
 			return ctrl.Result{Requeue: true}, err
 		}
-		return ctrl.Result{}, nil
+		nodeMetric = &slov1alpha1.NodeMetric{}
```
Comment (jasonliu747) [inline pkg/slo-controller/noderesource/noderesource_controller.go:92]: Do we need to reset NodeMetric here? 
Please correct me if I'm wrong. I thought `nodeMetric` would remain empty if there is an error when calling `Get()`.

File: pkg/slo-controller/noderesource/noderesource_controller_test.go [hunk 2]
```diff
@@ -18,8 +18,8 @@ package noderesource
 
 import (
 	"context"
+
```
Comment (jasonliu747) [inline pkg/slo-controller/noderesource/noderesource_controller_test.go:21]: no need to add a new line between `context` and `testing`

File: pkg/slo-controller/noderesource/noderesource_controller_test.go [hunk 3]
```diff
@@ -197,20 +198,15 @@ func Test_NodeResourceController_ColocationEnabled(t *testing.T) {
 	assert.Equal(t, int64(65000), batchCPU)
 
 	// reset node resources
-	r.Clock = clock.NewFakeClock(r.Clock.Now().Add(time.Hour))
```
Comment (jasonliu747) [inline pkg/slo-controller/noderesource/noderesource_controller_test.go:200]: will `isDegradeNeeded` be covered after this fake clock got deleted?

Comment (jasonliu747) [reply pkg/slo-controller/noderesource/noderesource_controller_test.go:200]: never mind. I saw you add a test for `isDegradeNeeded` exclusively in `pkg/slo-controller/noderesource/plugins/batchresource/plugin_test.go`

File: pkg/slo-controller/noderesource/resource_calculator.go [hunk 4]
```diff
@@ -120,7 +104,8 @@ func (r *NodeResourceReconciler) updateNodeResource(node *corev1.Node, nr *frame
 func (r *NodeResourceReconciler) updateNodeExtensions(node *corev1.Node, nodeMetric *slov1alpha1.NodeMetric, podList *corev1.PodList) error {
 	// update device resources
 	if err := r.updateDeviceResources(node); err != nil {
```
Comment (jasonliu747) [inline pkg/slo-controller/noderesource/resource_calculator.go:106]: should we add a config or feature-gate for device calculation?

File: pkg/slo-controller/noderesource/resource_calculator.go [hunk 5]
```diff
@@ -120,7 +104,8 @@ func (r *NodeResourceReconciler) updateNodeResource(node *corev1.Node, nr *frame
...
-		klog.Errorf("failed to update device resources for node %s, err: %v", node.Name, err)
+		klog.V(4).InfoS("failed to update device resources for node", "node", node.Name,
```
Comment (jasonliu747) [inline pkg/slo-controller/noderesource/resource_calculator.go:107]: why use InfoS here?

Comment (saintube) [reply pkg/slo-controller/noderesource/resource_calculator.go:107]: The `Reconcile` will log the error of `updateNodeExtensions`, so here it just log for `updateDeviceResources` with a higher verbosity.

File: pkg/slo-controller/noderesource/resource_calculator.go (hunk 4 above)
Comment (saintube) [reply pkg/slo-controller/noderesource/resource_calculator.go:106]: It has been discussed in https://github.com/koordinator-sh/koordinator/pull/1065#issuecomment-1457451506, so we do not change the device calculation in the issue #1060.

File: pkg/slo-controller/noderesource/noderesource_controller.go (hunk 1 above)
Comment (saintube) [reply pkg/slo-controller/noderesource/noderesource_controller.go:92]: OK. The nodeMetric object here is got by `CacheReader`, so it should remain empty when an error is returned.

File: pkg/slo-controller/noderesource/noderesource_controller_test.go (hunk 2 above)
Comment (saintube) [reply pkg/slo-controller/noderesource/noderesource_controller_test.go]: nice catch. fixed

File: pkg/slo-controller/noderesource/framework/extender_plugin.go [hunk 6]
```diff
@@ -17,122 +17,138 @@ limitations under the License.
...
+	globalNodePrepareExtender       = NewRegistry("NodePrepare")
+	globalNodeSyncExtender          = NewRegistry("NodeSync")
+	globalResourceCalculateExtender = NewRegistry("ResourceCalculate")
 )
 
+// Plugin has its name and priority. Registered plugins are executed in order of the priority. The high-priority plugin
+// runs earlier than the low-priority plugin. If the priorities are equal, compare using the lexicographic order.
+type Plugin interface {
+	Name() string
+	Priority() int32
```
Comment (zwzhang0107) [inline pkg/slo-controller/noderesource/framework/extender_plugin.go:38]: consider naming like sequence or something

File: pkg/slo-controller/noderesource/plugins/batchresource/plugin.go [hunk 7]
```diff
@@ -40,12 +42,18 @@ const PluginName = "BatchResource"
...
+var Clock clock.Clock = clock.RealClock{} // for testing
+
 type Plugin struct{}
 
 func (p *Plugin) Name() string {
 	return PluginName
 }
 
+func (p *Plugin) Priority() int32 {
+	return framework.DefaultPluginPriority
```
Comment (zwzhang0107) [inline pkg/slo-controller/noderesource/plugins/batchresource/plugin.go:54]: priority/seq can be overwrite in local repo, maybe defined all seq in an independent file.

File: pkg/slo-controller/noderesource/framework/extender_plugin.go (hunk 6 above)
Comment (saintube) [reply pkg/slo-controller/noderesource/framework/extender_plugin.go]: removed. just use the order of registration

File: pkg/slo-controller/noderesource/plugins/batchresource/plugin.go (hunk 7 above)
Comment (saintube) [reply pkg/slo-controller/noderesource/plugins/batchresource/plugin.go]: done. the ordering and registration are placed in `/noderesource/plugins_profile.go`

**Comment by hormes** [comment]
Setting the priority on the plugin will make understanding very complicated, because the logic is scattered in different files, it is difficult to add a plugin to know where it should be placed. If there is a strong order dependency between plugins, then we should have a place to register each plugin uniformly and arrange their order here.

**Comment by hormes** [comment]
I see that priority has been removed, and some comments and functions have not been removed

**Comment by saintube** [comment]
> I see that priority has been removed, and some comments and functions have not been removed

Fixed.

File: pkg/slo-controller/noderesource/framework/registry_test.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+		r.Unregister(p0.Name())
+		assert.Equal(t, wantPlugins1, r.GetAll())
+	})
+}
+
+var _ Plugin = (*testPlugin)(nil)
+
+type testPlugin struct {
+	name     string
+	priority int32
```
Comment (hormes) [inline pkg/slo-controller/noderesource/framework/registry_test.go:52]: the priority field is no longer valid

Comment (saintube) [reply pkg/slo-controller/noderesource/framework/registry_test.go]: removed

**Review by jasonliu747 — COMMENTED** [review]
/lgtm
others PTAL @hormes @zwzhang0107

**Pull request merged**